File: DataBaseRoomApp/jroomapp/src/main/java/com/woowrale/jroomapp/data/datasource/ContactDataSource.java
package com.woowrale.jroomapp.data.datasource;

import com.woowrale.jroomapp.data.database.ContactDao;
import com.woowrale.jroomapp.data.database.ContactDatabase;
import com.woowrale.jroomapp.data.mappers.MapperFactory;
import com.woowrale.jroomapp.domain.model.Contact;

import java.util.List;

public class ContactDataSource implements LocalDataSource {

    private ContactDao contactDao;

    public ContactDataSource(ContactDatabase contactDatabase) {
        contactDao = contactDatabase.contactDao();
    }

    @Override
    public Boolean isEmpty() {
        boolean isEmpty = true;
        if (contactDao.contactCount() > 0) {
            isEmpty = false;
        }
        return isEmpty;
    }

    @Override
    public void saveContacts(List<Contact> contacts) {
        contactDao.insertContact(MapperFactory.mapperContactsToContactEntity(contacts));
    }

    @Override
    public List<Contact> getPopularContacts() {
        return MapperFactory.mapperContactsEntityToContact(contactDao.getAll());
    }

    @Override
    public Contact findById(Integer id) {
        return MapperFactory.mapperContactEntityToContact(contactDao.findById(id));
    }

    @Override
    public void update(Contact contact) {
        contactDao.updateContact(MapperFactory.mapperContactToContactEntity(contact));
    }
}

File: DataBaseRoomApp/kroomapp/src/main/java/com/woowrale/kroomapp/data/datasource/ContactDataSource.kt
package com.woowrale.kroomapp.data.datasource

import com.woowrale.kroomapp.data.database.ContactDao
import com.woowrale.kroomapp.data.database.ContactDatabase
import com.woowrale.kroomapp.data.database.model.ContactEntity
import com.woowrale.kroomapp.data.mappers.toContact
import com.woowrale.kroomapp.data.mappers.toRoomContact
import com.woowrale.kroomapp.domain.model.Contact

class ContactDataSource (contactDatabase: ContactDatabase): LocalDataSource {

    private var contactDao: ContactDao = contactDatabase.contactDao()

    override fun isEmpty(): Boolean {return contactDao.contactCount() <= 0 }

    override fun saveContacts(contacts: List<Contact>) {
        contactDao.insertContact(contacts.map { it.toRoomContact() })
    }

    override fun getPopularContacts(): List<Contact> {
       return contactDao.getAll().map { it.toContact() }
    }

    override fun findById(id: Int): Contact {
        return contactDao.findById(id).toContact()
    }

    override fun update(contact: Contact) {
        contactDao.updateContact(contact.toRoomContact())
    }
}